File: 01/node/part_one.js
const { lineReader, getNextLine } = require("./lineReader");

(async () => {
  const reader = lineReader("input.txt");

  let leftList = [];
  let rightList = [];
  let line;
  while ((line = await getNextLine(reader)) !== null) {
    const locations = line
      .split(" ")
      .filter(Boolean)
      .map((n) => parseInt(n));

    leftList.push(locations[0]);
    rightList.push(locations[1]);
  }

  leftList.sort((a, b) => a - b);
  rightList.sort((a, b) => a - b);

  let distance = 0;
  for (let i = 0; i < leftList.length; i++) {
    distance += Math.abs(leftList[i] - rightList[i]);
  }

  console.log(`Distance: ${distance}`); // 1765812
})();

File: 01/c++/part_one.cpp
#include <iostream>
#include <vector>
#include <sstream>
#include <algorithm>
#include "LineReader.hpp"

std::vector<int> parseLine(const std::string& line) {
    std::vector<int> numbers;
    std::stringstream ss(line);
    int number;

    while (ss >> number) {
        numbers.push_back(number);
    }

    return numbers;
}

/**
 * g++ -o main LineReader.cpp main.cpp
 */
int main() {
    try {
        LineReader reader("input.txt");
        std::vector<int> leftList;
        std::vector<int> rightList;

        while (true) {
            auto line = reader.getNextLine();
            if (!line.has_value()) {
                break; 
            }
            std::cout << "Read line: " << line.value() << std::endl;

            auto numbers = parseLine(line.value());

            leftList.push_back(numbers[0]);
            rightList.push_back(numbers[1]);
        }

        std::sort(leftList.begin(), leftList.end());
        std::sort(rightList.begin(), rightList.end());

        int distance = 0;
        for (int i = 0; i < leftList.size(); i++) {
            distance += abs(leftList[i] - rightList[i]);
        }

        std::cout << distance << std::endl;

    } catch (const std::exception& e) {
        std::cerr << "Error: " << e.what() << std::endl;
        return 1;
    }

    return 0;
}
